### DISCUSSION TITLE
Is HTTP PATCH supported?

### DISCUSSION THREAD

**Discussion by JohnD832** [serenity-is/Serenity#6862] [General]
Hi all

I was wondering, is HTTP PATCH supported in Serenity? This is useful when doing an Update to an entity.

**Comment by volkanceylan** [comment]
Serenity entities and json handling is smart enough to only update fields passed in json, no patch required

**Comment by JohnD832** [comment]
Thanks for getting back to me @volkanceylan.

In our system, we have exposed access to our XYZRow.cs entities through an API. The API is a separate project that references a lot of the code in the serenity project.

Here's a typical API endpoint:

```
[HttpPut("{id}")]
[Produces("application/json")]
public async Task<ActionResult> Update(int id, [FromBody] XYZ entity)
{
    try
    {
        if (!ModelState.IsValid)
        {
            return BadRequest(ModelState);
        }

        entity.Id = id;
        var json = JsonConvert.SerializeObject(entity);
        var param = new KeyValuePair<string, string>("JsonRow", json);

        GetLoginCredential(Request);
        KeyValuePair<string, string>[] parameter = { param, impersonateAccount };
        var isUserAuthenticate = await LoginHelper.AuthenticateUser(bindApi, serviceUrl, username, password);
        if (!isUserAuthenticate)
        {
            return BadRequest("Authentication Error");
        }
        var response = await bindApi.UpdateRequest<XYZ>(serviceUrl, baseUrl, parameter);

        if (response.Error == null)
        {
            return Ok(response);
        }
        else
        {
            if (response.Error.Code == notFoundResult.StatusCode.ToString())
                return NotFound();

            return BadRequest(response.Error);
        }
    }
    catch (Exception ex)
    {
        return BadRequest($"Unable to update the record. Message: {ex.Message}");
    }
}
```

I think we might be doing this wrong. Is there a recommended way of creating a CRUD API around our all our XYZRow.cs files?

**Reply by VictorTomaili** [reply]
Sergen automatically creates the crud endpoint in serenity application. You can also create endpoints manually. Check the serene template and northwind sample. 
https://github.com/serenity-is/common-features/blob/master/src/Serenity.Demo.Northwind/Modules/Category/CategoryEndpoint.cs

**Reply by JohnD832** [reply]
Thanks @VictorTomaili, so you're saying there's no need for a separate project? We just add **JSON returning** endpoints to the existing `XYZEndpoint.cs` files? For example, we might want an endpoint to get a list of XYZRow.cs as JSON? Or an endpoint to the update the entity with a PATCH request?

**Comment by VictorTomaili** [comment]
XYZEndpoint.cs already contains update and list functionality when generated. And they return the data in JSON format. A PATCH request is not necessary. 
Here is a sample of how serenity endpoints works. 

This is a create request. All the fields are filled. 
```json
{"Entity":{"CategoryName":"Test Category","Description":"This is a test category"}}
```

This is an update request. Only the Description is filled. The only description field is updated. 
```json
{"Entity":{"Description":"This is a test category [update]"},"EntityId":11}
```

This is an update which sets the description to null
```json
{"Entity":{"Description": null},"EntityId":11}
```

The JSON is capable to understand which values are existing in request. And serenity entities are clever objects which can handle null and not existing fields. The PATCH request is not needed. We did not use RESTFUL endpoints. The serenity uses an easier to use API style. Just send the JSON object with fields which you want to be updated.

**Reply by JohnD832** [reply]
Thanks @VictorTomaili. That's very interesting, I'm learning a lot here. I have a question. Right now, in our app when I visit `/organisation/xyz`, I get shown the `XYZIndex.cshtml` page. What URL do I need to go to in order to get the `JSON` for a list of XYZ entities?

**Reply by dfaruque** [reply]
`services/organisation/xyz/List`

**Reply by JohnD832** [reply]
Thanks @dfaruque <3
